src/wasm-lib/kcl/src/execution/memory.rs

```diff
@@ -1,74 +1,405 @@
+//!
+//! - The current env and all envs in the call stack are 'just envs', never a snapshot (we could
+//!   use just a ref to an env, rather than to a snapshot but this is pretty inconvenient, so just
+//!   know that the snapshot ref is always to the current version). Only the parent envs or saved refs
+//!   can be refs to snapshots.
+//! - We only ever write into the current env, never into any parent envs (though we can read from
+//!   both).
+//! - Therefore, there is no concept of writing into a snapshot, only reading from one.
+//! - The env ref saved with a function decl is always to a snapshot, never to the current version.
+//! - If there are no snapshots in an environment and it is no longer in the call stack, then there
+//!   are no references from function decls to the env (if it is the parent of an env with extant refs
+//!   then there would be snapshots in the child env and that implies there must be a snapshot in the
+//!   parent to be the parent of that snapshot).
+//! - Since KCL does not have submodules and decls are not visible outside of a nested scope, all
+//!   references to variables in other modules must be in the root scope of a module.
+//! - Therefore, an active env must either be on the call stack, have snapshots, or be a root env. This
+//!   is however a conservative approximation since snapshots may exist even if there are no live
+//!   references to an env.
+
+use std::{collections::HashMap, fmt};
+
 use anyhow::Result;
 use indexmap::IndexMap;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 
 use crate::{
     errors::{KclError, KclErrorDetails},
-    execution::{KclValue, Metadata, Sketch, Solid, TagIdentifier},
+    execution::{KclValue, Metadata},
     source_range::SourceRange,
 };
+use env::Environment;
 
-#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, ts_rs::TS, JsonSchema)]
+/// The distinguished name of the return value of a function.
+pub(crate) const RETURN_NAME: &str = "__return";
+
+/// KCL memory. There should be only one ProgramMemory for the interpretation of a program (
+/// including other modules). Multiple interpretation runs should have fresh instances.
+///
+/// See module docs.
+#[derive(Debug, Clone, Deserialize, Serialize, ts_rs::TS, JsonSchema)]
 #[ts(export)]
 #[serde(rename_all = "camelCase")]
 pub struct ProgramMemory {
-    pub environments: Vec<Environment>,
-    pub current_env: EnvironmentRef,
-    #[serde(rename = "return")]
-    pub return_: Option<KclValue>,
+    environments: Vec<Environment>,
+    /// Invariant: current_env.1.is_none()
+    current_env: EnvironmentRef,
+    /// Invariant: forall er in call_stack: er.1.is_none()
+    call_stack: Vec<EnvironmentRef>,
+    /// Statistics about the memory, should not be used for anything other than meta-info.
+    #[allow(dead_code)]
+    #[serde(skip)]
+    pub(crate) stats: MemoryStats,
+}
+
+// Intended for debugging. Do not rely on this output in any way!
```

Comment:
Why not impl Debug instead then?

Reply:
It's a very cut-down display of the memory and I figured we might want the generated debug as well as this short version